### PR TITLE
add recurssive merge `rec_merge`

### PR DESCRIPTION
**Pull request by wytbella**
Add function `rec_merge` to recursively merge namedtuple of namedtuple of namedtuple...

### CODE REVIEW

File: src/NamedTupleTools.jl [hunk 1]
```diff
@@ -27,6 +28,8 @@ else
      export fieldtypes
 end
 
+struct NotPresent end
```
Comment (oxinabox) [inline src/NamedTupleTools.jl:31]: Maybe a comment  explaining  what this is for

File: src/NamedTupleTools.jl [hunk 2]
```diff
@@ -298,6 +301,33 @@ merge(a::NamedTuple{an}, b::NamedTuple{bn}, c::NamedTuple{cn}, d::NamedTuple{dn}
...
     reduce(merge,(a, b, c, d, e, f, g))
 
+"""
+    rec_merge(nt1, nt2)
+    rec_merge(nt1, nt2, nt3, ..)
+
+Recurssively merge namedtuples. Fieldnames and values in nt2 and its sub-namedtuples
+    are all kept, with those only appear in nt1.
+
+see: [`delete!`](@ref)
```
Comment (oxinabox) [inline src/NamedTupleTools.jl:311]:
```suggestion
see: [`merge`](@ref)
```

File: src/NamedTupleTools.jl (hunk 1 above)
Comment (wytbella) [reply src/NamedTupleTools.jl:36]: added

**Comment by JeffreySarnoff** [comment]
@wytbella  This is a very good idea. Thank you. I am reviewing the details.

**Comment by JeffreySarnoff** [comment]
I think I am missing something (or maybe I am testing on a slightly different version of NamedTupleTools).  The tests for `rec_merge` pass using `merge` instead of `rec_merge` (see below).  Would you show me some cases where they differ?
```
julia> nt1
(a = 1, b = 2, c = 3, d = 4)

julia> nt2
(a = "one", b = "two")

julia> ntnt1
(x = (a = 1, b = 2, c = 3, d = 4), y = (a = "one", b = "two"))

julia> ntnt2
(y = (a = 1.0, b = "two"),)

julia> @test rec_merge(nt1) == nt1
Test Passed

julia> @test rec_merge(nt1, nt2) === (a = "one", b  = "two", c = 3, d = 4)
Test Passed

julia> @test rec_merge(nt1, nt2, nt1) == (a = 1, b  = 2, c = 3, d = 4)
Test Passed

julia> @test rec_merge(ntnt1) == ntnt1
Test Passed

julia> @test rec_merge(ntnt1, ntnt2) == (
           x = (a = 1, b = 2, c = 3, d = 4),
           y = (a = 1.0, b = "two")
       )
Test Passed

julia> @test rec_merge(ntnt1, ntnt2, ntnt1) == (
           x = (a = 1, b = 2, c = 3, d = 4),
           y = (a = "one", b = "two")
       )
Test Passed

julia> @test merge(nt1) == nt1
Test Passed

julia> @test merge(nt1, nt2) === (a = "one", b  = "two", c = 3, d = 4)
Test Passed

julia> @test merge(nt1, nt2, nt1) == (a = 1, b  = 2, c = 3, d = 4)
Test Passed

julia> @test merge(ntnt1) == ntnt1
Test Passed

julia> @test merge(ntnt1, ntnt2) == (
           x = (a = 1, b = 2, c = 3, d = 4),
           y = (a = 1.0, b = "two")
       )
Test Passed

julia> @test merge(ntnt1, ntnt2, ntnt1) == (
           x = (a = 1, b = 2, c = 3, d = 4),
           y = (a = "one", b = "two")
       )
Test Passed
```

**Comment by wytbella** [comment]
Good point! Thansk for pointing that out.

The test examples are not good to distinguish the difference between `merge` and `rec_merge`. Please check the updated tests. Basically,
- for single layer namedtuple (e.g. nt1, nt2), merge and rec_merge are expected to bahave in the same way.
- for multi-layer namedtuple (e.g. ntnt1, ntnt2), they are expected to behave differently. The different bahaviours are now tested.

**Comment by JeffreySarnoff** [comment]
Thank you for the update. 

> for multi-layer namedtuple (e.g. ntnt1, ntnt2), they are expected to behave differently. 

Please tell me in words how rec_merge is expected to behave given a "multi-layer namedtuple".

**Comment by wytbella** [comment]
> Please tell me in words how rec_merge is expected to behave given a "multi-layer namedtuple".

say `a` and `b` are both 3-layer namedtuple. `rec_merge(a,b)` goes layer by layer to apply a normal merge. In contrast, `merge(a,b)` will do it on the most outside layer.

**Comment by JeffreySarnoff** [comment]
Please show me an example: three-layer `a`, three-layer `b`, merge(a,b), rec_merge(a,b).  I am not clear about whether three-layer means 3 single nestings or a 3 level nesting:`a=(x=(p=1,q=2), y=(..), z=(..))` or `a=(x=(p=1, q=(r=(s=2, t=3)))`; how do you want multi-level nestings to be handled?

**Comment by wytbella** [comment]
3-layer nesting means 3 levels. Here is an example

```julia
julia> using NamedTupleTools

julia> a = (;x=(p=1, q=(;r=(s=2, t=3))))
(x = (p = 1, q = (r = (s = 2, t = 3),)),)

julia> b = (;x=(; q=(;r=(;s="two"))))
(x = (q = (r = (s = "two",),),),)

julia> merge(a,b)
(x = (q = (r = (s = "two",),),),)

julia> rec_merge(a,b)
(x = (p = 1, q = (r = (s = "two", t = 3),)),)
```

**Comment by JeffreySarnoff** [comment]
is `rec_merge(b,a) == (x = (p = 1, q = (r = (s = 2, t = 3),)),)`?

**Comment by wytbella** [comment]
yes

```julia
julia> rec_merge(b,a)
(x = (q = (r = (s = 2, t = 3),), p = 1),)
```

**Comment by wytbella** [comment]
Hi @JeffreySarnoff, do you have any other comments/questions on this PR?

**Comment by JeffreySarnoff** [comment]
I have been under the weather.  This is not forgotten, just delayed.

**Comment by JeffreySarnoff** [comment]
apologies for the delay -- I will resolve this over the weekend.
It would help me to have two examples where it is useful.

**Comment by JeffreySarnoff** [comment]
I would prefer not introducing `NotPresent`.  It seems to me that we could accomplish the same thing using `Nothing` in place of `NotPresent` and using `nothing` in place of `NotPresent()`.  Is there some reason that it is necessary to introduce `NotPresent`?

**Comment by JeffreySarnoff** [comment]
@wytbella I am incorporating your algorithm and implementation without directly merging.  This allows me to make two modifications without asking you to do it:
(1) rather than `merge_rec` as a separate function, I am adding a keyword `recursive::Bool=false` to `merge` and selecting your work when `recursive=true`.
(2) rather than introduce `NotPresent` internally, I am using `Missing, missing`.

(1) seems cleaner and easier for people to remember, for the user it should work the same as `rec_merge` does. The only implementation distinction is (2).  I tried some test cases and I have not encountered any difference.  If you have a corner case that I have missed, please let me know.

The purpose of (2) is to stay within standard Julia exception and recursion processing [read "prefer not to introduce new sentinel structs where there is an available alternative that works"], otherwise each package that implements special structure recursion would need its own internal sentinel and that gets messy.

I have to copy the revised code (see branch `test` for a test of (2)), add a couple of examples and bump the version and release this.  It is all underway. 

Please see the `README.md` in current master; your contribution is acknowledged.
It is well earned, and it took me too long to catch up with your core intent.
I am appreciative of the thought, the work, the cleanups, and your patience.

Best, Jeffrey

**Comment by JeffreySarnoff** [comment]
To avoid problematic type-piracy, I could not add a keyword to `merge`. The recursive version of the merge function is named `merge_recursive`. Your work has been merged and the version is updated. Registration of the new version is pending.

**Comment by wytbella** [comment]
> I would prefer not introducing NotPresent. It seems to me that we could accomplish the same thing using Nothing in place of NotPresent and using nothing in place of NotPresent(). Is there some reason that it is necessary to introduce NotPresent?

The concern of using `missing` or `nothing` instead is that what bout `missing/nothing` is actually the value you want to merge. For example if `a = (;x=1)` and `b=(;x=missing)`. When I do `rec_merge(a,b)`, I might want to get `(;x=missing)` back rather than `(;x=1)`

**Comment by JeffreySarnoff** [comment]
Yes, now I understand.  I will fix that.
